### DISCUSSION TITLE
How to 'transfer' or import users?

### DISCUSSION THREAD

**Discussion by cvrle77** [muraliwebworld/push-notification-plugin-for-wordpress-websites#7] [Q&A]
Since I've been using some other plugins (Perfecty Push Notifications) which is now abandoned, a lot of my visitors are already subscribed to my notifications.
That creates an issue, because I can't import them, and I can't reset notification subscription in their browser.
Popup that ask users to subscribe doesn't show up, unless they manually reset their subscription.
Is there a way to handle this issue in some manner?
Could it be possible to send notifications to users, and once they are on the website, to update or reset their subsciption, so they can subscribe with new service?

**Comment by muraliwebworld** [comment]
Perfecty push notification is based on web push (self hosted). This plugin is designed using Firebase as push notification provider. Difference between web push and Firebase is as follows,

Web push creates subscription from browser which consists of endpoint, vapid key for every subscription there is no intermediate token
In Firebase, once user subscribed from browser, this plugin will communicate with Firebase to create intermediate token for every subscription and token will be stored in Firebase as well as in this plugin’s PNFPB database table. Notifications will be sent only to tokens. In latest httpv1 Firebase it is based on topics. Firebase will create vapidkey from the while creating the project which consists of web app, android/ios app credentials which will be used by this plugin to send notifications.
Inorder To migrate from perfecty push notifications,once this plugin is installed and activated after deactivating perfecty, when user visits your site from their browser, if browser is already subscribed this plugin will communicate with firebase to create token, it will be stored in backend database (this plugin database). In order to setup this plugin, 1. create Firebase account 2. Create Firebase project 3. Create web app 4. Use those credentials to setup this plugin as instructed in Firebase configuration mentioned in admin area/plugin installation details .
If you need more details, please contact murali@indiacitys.com support id mentioned in admin area or in github library https://github.com/muraliwebworld/push-notification-plugin-for-wordpress-websites. It will be better to create test version of your site first by cloning data from your live site or with some limited users from your live site to test this. Migration will not require any special coding as Firebase creates token for every subscribed request using Firebase api.

**Comment by cvrle77** [comment]
If I understood correctly, all Perfecty subscribers should be migrated to PNFPB automatically when they land on my website?

For some reason, in 'real life', it doesn't work. I've tried it myself, on my own website.

If I am subscribed with Perfecty Push, and I disable Perfecty Push, my browser still shows that I am subscribed to notifications.
When I turn on PNFPB, land on my website, I am not getting token, I am being prompted to subscribe to notifications in new plugin, even though I can see I am already subscribed (through Perfecty)

I've tested, and both plugins can work in parallels, both show subscription message at once, both workers work, and each plugin can send push message to it's base of subscribers. But nothing as described happens no token creation happens.

![image](https://github.com/muraliwebworld/push-notification-plugin-for-wordpress-websites/assets/16167396/1d7e511a-5938-49ca-ad38-7720bd79de5a)

**Comment by muraliwebworld** [comment]
I will test this in one of my site and I will let you know

**Comment by muraliwebworld** [comment]
I have tried in my site muraliwebworld.com in following sequence,
1. I have cleared/deleted all tokens in PNFPB plugin under device tokens tab
2. I deactivated PNFPB plugin
3. I installed perfecty push and activated
4. Both service workers are active
5. I sent notification from admin area of perfect push 
6. I got only one notification from perfecty eventhough both service workers are present. Even both service workers are in same scope, it took perfecty plugin service worker. If you have any other cache plugin, please clear the cache while swapping the plugins. 
7. I deactivated perfecty and activated pnfpb, came to front end, device token got created automatically and sent one time push notification
8. I got only pnfpb notification. Following are screenshots,
(If you have any cache plugin like yousify or wp super cache or wp total cache or any other minify cache plugin, please make sure clear cache of site from admin area while swapping the plugins with service workers, so that it will take correct service worker when frontend user visited the site even with multiple service workers registered)
I think in your site, you are using Redis Object Cache (i think it is from hosting side), please clear out that cache while swapping between plugins
**Following are screenshots,**
![Screenshot 2024-05-16 at 12 26 00 PM](https://github.com/muraliwebworld/push-notification-plugin-for-wordpress-websites/assets/32461311/3fa72252-d09a-403d-8598-9bf1e522de6e)
![Screenshot 2024-05-16 at 12 26 43 PM](https://github.com/muraliwebworld/push-notification-plugin-for-wordpress-websites/assets/32461311/46f4e86c-168c-4ce0-bea7-3c2a8f54ef9c)
![Screenshot 2024-05-16 at 12 27 46 PM](https://github.com/muraliwebworld/push-notification-plugin-for-wordpress-websites/assets/32461311/f7a157e6-db10-4f6d-b5e9-e6347529989f)
![Screenshot 2024-05-16 at 12 28 01 PM](https://github.com/muraliwebworld/push-notification-plugin-for-wordpress-websites/assets/32461311/1e5b4c15-2c7c-4f2a-b584-01244debb1cb)
![Screenshot 2024-05-16 at 12 29 12 PM](https://github.com/muraliwebworld/push-notification-plugin-for-wordpress-websites/assets/32461311/8b0b8495-c381-40e1-a5e5-36d36a0824ed)

**Comment by muraliwebworld** [comment]
In addition to my previous reply, I think in your site, you are using Redis Object Cache (i think it is from hosting side), please clear out that cache in your hosting side while swapping between plugins, in below screenshots, it shows disk cache while reloading your site, it is due to redis object cache.
![Screenshot 2024-05-16 at 12 53 05 PM](https://github.com/muraliwebworld/push-notification-plugin-for-wordpress-websites/assets/32461311/3c0f99f7-e9df-4b76-a345-975b1f167450)

**Comment by cvrle77** [comment]
Ok, so I've wanted to test this on localhost, and I've installed Laragon, without Redis, and NGINX, bare apache+mysql.
I found a culprit.
It's the option 'Show custom prompt to subscribe push notification.'

If the option is enabled, then custom prompt is shown regardless of my previous subscription to Perfecty, and token isn't given.
If the option is disabled, token is given, and only default browser prompt for notification subscription is shown to user.

Not sure if feature or a bug, but you wrote: '' It is mandatory, if safari browser is used for push notification and also custom prompt is needed for Firefox android''

So, it would be nice to have both custom prompt for those that aren't subscribed, and token, for those that are subscribed, at the same time.

![image](https://github.com/muraliwebworld/push-notification-plugin-for-wordpress-websites/assets/16167396/7c90b093-2b1a-40ff-99e7-0a571ada8b33)

**Comment by muraliwebworld** [comment]
If custom prompt is enabled and it is subscribed to perfecty it will show only unsubscribe button as it is already subscribed (if you click on bell prompt it will show unsubscribe button as it is already subscribed). After enabling pnfpb plugin, is it possible to send me screenshot of your site with dev tools console tab messages, i want to see any other warning or errors are present in your site. Firebase will not take browse subscription, only pnfpb plugin will check whether browser is subscribed for notifications, once it is subscribed (irrespective of any plugins), pnfpb plugin will make api call to firebase to create token, this token will be stored in plugin's table and used to send notifications. If you have any test version of your site, it will be easy to check and it will be easy for me to help you. For my site, it is working fine when i swap between perfecty or pnfpb or onesignal or some other plugins.

**Comment by cvrle77** [comment]
I have to write this, because I can't understand.

I went to a test site, did everything like I was doing previously, and I couldn't make it work properly.

but the only thing I did differently once, is to enable BELL prompt option in custom prompt.

Then everything started working on test site. I've tried all possible combinations with prompt styles, with or without bell, and every time I would get token. I've turned on and off the plugin, switching from one to another, keeping them both, it would work. For some reason, everything got stuck because of that option.

Now I went to my working website to test this hypothesis, I've tested without bell, and it still didn't work. Once I've turned on bell prompt, I got token, and I also got token in every other combination. 
I am running Perfecty, and I've disabled showing up notification completely. 
In parallel, I am running PNFPB.

I am using Perfecty to send notifications to my old subs, and once they land on my site, PNFPB is 'catching' them and giving them token, transferring them to PNFPB.

I've no idea why it got stuck, but it's finally unstuck.

**But I have to ask - what is the difference between Userid 1 and Userid 0? And what is the meaning of "Shortcode subscription 1000000000"**

**Comment by cvrle77** [comment]
Actually it's just a matter of turning on and off custom prompt and saving in between the two stages to reactivate it.
It's like when you turn on the plugin while that option is turned on, tokenization is asleep. Only when you turn off/turn on (and save) custom prompt, it wakes up and starts working.

Userid 1 - logged in user
Userid 0 - not logged in
Still haven't figured out what is  Shortcode subscription 1000000000

**Reply by muraliwebworld** [reply]
Push notification can be subscribed even without logging in. If user is not logged in but subscribed to push notification then user id will be zero only token will be present. If user subscribed after logging or if user logged in after subscribing (before login) then user id will be updated to corresponding token. Subscription option is different, please see my last reply which is sent before this reply below update

**Comment by muraliwebworld** [comment]
Shortcode subscription 1000000000 indicates all options like post/activities/groups/friend request/accept extra are subscribed. If you enable frontend subscription option (if buddypress is installed) then you will be able to change in front end, it will be applicable only for BuddyPress users as most of options are related to buddypress like activities, friend request/accept/ private messages. First byte indicates are all subscribed, second one for post, third one for buddypress all comments, fourth one for buddypress comments only to author, fifth one for private messages, sixth byte for new member added, seventh byte for friend request, eight byte for friend accept, ninth byte for buddypress user avatar change, tenth byte for buddypress cover image change, 10th byte for unsubscribe all options, 12th byte for buddypress group details update, 13th byte for buddypress group invite update. In the year 2019 during first release of this plugin, i designed this plugin only with post and buddypress activities and then i added one by one option every release till now. If you do not have buddypress then if first byte indicates 1 means then you are subscribed to post and if 10th byte indicates 1 then you are unsubscribed to all options. 1 means subscribed and 0 means option is unsubscribed or subscribed to other options. If you do not have buddypress you no need to worry about this option, if you unsubscribe it will unsubscribe notifications from post. If you have buddypress, then enable frontend notification settings in **frontend settings tab in admin area** and frontend users can change their options **like in below screenshots - (in Frontend) buddypress user account -> settings-> push notification subscription menu** , accordingly it will update subscription options like in below screenshots, in below screenshot, i **switched off** avatar change and cover image change, then subscription option in device tokens will be 01011111000111 or you can use shortcode subscribe_PNFPB_push_notification to display/control frontend notifications for front end users
![Screenshot 2024-05-20 at 6 02 47 PM](https://github.com/muraliwebworld/push-notification-plugin-for-wordpress-websites/assets/32461311/24dbfd96-765b-4cc8-8ba4-1f0e7e1ceb28)
![Screenshot 2024-05-20 at 6 03 08 PM](https://github.com/muraliwebworld/push-notification-plugin-for-wordpress-websites/assets/32461311/95e8d5a7-cc25-4dff-adf5-ba3c78a0b895)

**Comment by cvrle77** [comment]
Your effort to explain something, is on another level. Thank you, thank you, thank you.
